Automattic/wp-calypso
# chore: Extract babel config to a separate package

## Background

- We only have one config file for the whole repo. This forces individual packages to import babel plugins used by that root config (which confuses tools like [depcheck](https://www.npmjs.com/package/depcheck)), or assume that somebody else will install them in `node_modules` (which breaks yarn [Focused Workspaces](https://yarnpkg.com/cli/workspaces/focus/#gatsby-focus-wrapper)).

 - Other projects use some of our Babel presets, but they are exposed as part of `@automattic/calypso-build`. This forces consumers to bring in the whole package, which requires some beefy peer dependencies (cc @Automattic/jetpack-crew)


## Changes

Extracts Babel config to its own package, `@automattic/calypso-babel-config`.

This package provides both the main config used in Calypso (i.e. the content of `babel.config.js`). This allows us to have a single place to centralize babel config and dependencies, instead of having them scattered across multiple projects.

This new package also provides the same presets that were available in `@automattic/calypso-build/presets/*`. This should be a drop-in replacement if the consumer is not interested in other parts of `@automattic/calypso-build`

## Testing instructions

Build each app and verify the output has not changed. For each app I listed the command to build it, and the command to diff the output with `trunk`. It assumes you have a checkout of `trunk` in `wp-calypso-trunk` and a checkout of this branch in `wp-calypso`.

Note: This assumes that the PR is up-to-date with `trunk`. If it is not, you can use the commit 0104597264 to compare against, as it is the most recent common ancestor.

### Calypso

* Build command: `NODE_ENV=production CALYPSO_ENV=production BUILD_TRANSLATION_CHUNKS=true yarn build`
* Diff command: `diff -qr wp-calypso/public/evergreen wp-calypso-trunk/public/evergreen`

The only difference should be in `manifest.*.js`, as it contains the commit hash:

```
$ diff -qr wp-calypso/public/evergreen wp-calypso-trunk/public/evergreen
Only in wp-calypso-trunk/public/evergreen: runtime.47abd30ceb27f3555d30.min.js
Only in wp-calypso/public/evergreen: runtime.6845260d58ce2a58eda7.min.js
```

### ETK

* Build command: `cd apps/editing-toolkit && yarn build`
* Diff command: `diff -qr wp-calypso/apps/editing-toolkit/editing-toolkit-plugin wp-calypso-trunk/apps/editing-toolkit/editing-toolkit-plugin`

The only difference should be in `gutenboarding-launch.asset.php`, as it contains the commit hash:

```
$ diff -qr wp-calypso/apps/editing-toolkit/editing-toolkit-plugin wp-calypso-trunk/apps/editing-toolkit/editing-toolkit-plugin
Files wp-calypso/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/dist/gutenboarding-launch.asset.php and wp-calypso-trunk/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/dist/gutenboarding-launch.asset.php differ
```

### Notifications

* Build command: `cd apps/notifications && yarn build` 
* Diff command: `diff -qr wp-calypso/apps/notifications/dist wp-calypso-trunk/apps/notifications/dist`.

There will be some differences in `index.html` and `rtl.html`, again related to the commit hash:

```
$ diff -qr wp-calypso/apps/notifications/dist wp-calypso-trunk/apps/notifications/dist
Files wp-calypso/apps/notifications/dist/cache-buster.txt and wp-calypso-trunk/apps/notifications/dist/cache-buster.txt differ
Files wp-calypso/apps/notifications/dist/index.html and wp-calypso-trunk/apps/notifications/dist/index.html differ
Files wp-calypso/apps/notifications/dist/rtl.html and wp-calypso-trunk/apps/notifications/dist/rtl.html differ
```

### O2 Blocks

* Build command: `cd apps/o2-blocks && yarn build`
* Diff command: `diff -qr wp-calypso/apps/o2-blocks/dist wp-calypso-trunk/apps/o2-blocks/dist`

There should be no differences.

### WBE

* Bulid command: `cd apps/wpcom-block-editor && yarn build`
* Diff command: `diff -qr wp-calypso/apps/wpcom-block-editor/dist wp-calypso-trunk/apps/wpcom-block-editor/dist`

There should be no differences.
